### PR TITLE
WIP: update timezones to track current PostgreSQL release

### PR DESCRIPTION
**Pull request by danielgustafsson**
WIP: while this passes ICW locally for me I'd like to see this exercised on the pipeline before I deem it ready.

The timezone data in Greenplum are from the base version of PostgreSQL that the current version of Greenplum is based on. This cause issues since it means we are years behind on tz changes that have happened. This pulls in the timezone data and code from PostgreSQL 10.1 with as few changes to Greenplum as possible to minimize merge conflicts. The goal is to gain data rather than features, and for Greenplum for each release to be able to stay current with the iana tz database as it is imported into upstream PostgreSQL.

This removes a Greenplum specific test for the Yakutsk timezone as it was made obsolete by upstream tz commit 1ac038c2c3f25f72. Further it backports the move to the zi dataformat which is much less verbose then the old tz files.

**Note:**  This PR depends on #4143 for fixing `pg_attribute_printf()`.

### CODE REVIEW

**Comment by Eulerizeit** [comment]
@schubert this is the timezone fix. It can certainly be backported to 5.x.

**Comment by danielgustafsson** [comment]
The careful observer will see that this does introduce a warning on pg_attribute_printf, that will naturally be fixed before I take the WIP label off.

**Comment by ivannovick** [comment]
+1 for backport, we should keep this updated regularly.

File: src/backend/utils/misc/tzparser.c [hunk 1]
```diff
@@ -124,7 +124,7 @@ splitTzLine(const char *filename, int lineno, char *line, tzEntry *tzentry)
 						filename, lineno)));
 		return false;
 	}
-	tzentry->abbrev = abbrev;
+	tzentry->abbrev = pstrdup(abbrev);
```
Comment (jchampio) [inline src/backend/utils/misc/tzparser.c]: The commit that added this strdup also removed a later strdup in `addToArray()`; does that need to be part of this backport?

File: src/timezone/Makefile [hunk 2]
```diff
@@ -58,6 +59,11 @@ ifeq (,$(with_system_tzdata))
 endif
 	$(MAKE) -C tznames $@
 
+abbrevs.txt: zic $(TZDATAFILES)
```
Comment (jchampio) [inline src/timezone/Makefile]: nitpick: Upstream has an additional entry in `.gitignore` for the new `abbrevs.txt` target.

File: src/include/datatype/timestamp.h [hunk 3]
```diff
@@ -0,0 +1,197 @@
...
+ * int64 values with units of microseconds.  (Once upon a time they were
+ * double values with units of seconds.)
+ *
+ * TimeOffset and fsec_t are convenience typedefs for temporary variables.
+ * Do not use fsec_t in values stored on-disk.
+ * Also, fsec_t is only meant for *fractional* seconds; beware of overflow
+ * if the value you need to store could be many seconds.
+ */
+
+typedef int64 Timestamp;
```
Comment (jchampio) [inline src/include/datatype/timestamp.h]: This file is duplicating several typedefs/defines from `utils/timestamp.h`. Do we want to backport more of a7801b62 to get rid of that?

File: src/backend/utils/misc/tzparser.c (hunk 1 above)
Comment (danielgustafsson) [reply src/backend/utils/misc/tzparser.c]: Nice catch, looking more at this part was on my mental todo but it slipped away it seems. Let me dive more into this.

File: src/include/datatype/timestamp.h (hunk 3 above)
Comment (danielgustafsson) [reply src/include/datatype/timestamp.h]: I started down that path but opted for keeping this single file just for `src/timezone` until we merge where it came from. I did actually at first keep this as `src/timezone/timestamp_gp.h` to really carve it out, but moved it back. Perhaps keeping it separate would be better in terms of keeping the tree clean?

File: src/timezone/Makefile (hunk 2 above)
Comment (danielgustafsson) [reply src/timezone/Makefile]: Will fix, thanks

**Comment by danielgustafsson** [comment]
> On 14 Dec 2017, at 13:52, Ivan Novick <notifications@github.com> wrote:
> 
> +1 for backport, we should keep this updated regularly.

The idea is that updating the timezone database should be part of the release procedure such that every release carries the latest available version.

File: src/include/datatype/timestamp.h (hunk 3 above)
Comment (jchampio) [reply src/include/datatype/timestamp.h]: Yeah, I'm not really sure which would be better. As long as no one but `src/timezone` actually *uses* the new file, it shouldn't be a problem (famous last words)...

File: src/backend/utils/misc/tzparser.c (hunk 1 above)
Comment (danielgustafsson) [reply src/backend/utils/misc/tzparser.c]: Looking at this it seems to me that the moved `pstrdup()` is related to `validateTzEntry()` clobbering on`abbrevs`, so owning the memory is a better move. I've removed the dup call in `addToArray()` to match that from upstream.

File: src/include/datatype/timestamp.h (hunk 3 above)
Comment (danielgustafsson) [reply src/include/datatype/timestamp.h]: I opted for moving the file and decorating with FIXME's, else we risk it being used when it shouldn't be used.

**Comment by danielgustafsson** [comment]
Updated with review comments and rebased on top of master to get the fix from #4143

**Comment by jchampio** [comment]
Looks good! For most of the files under `src/timezone` my review was just a diff against upstream and manual inspection, followed by running some of the timezone-related tests. If there's anything else you'd like me to check on my machine, let me know.

**Comment by danielgustafsson** [comment]
> On 15 Dec 2017, at 17:32, Jacob Champion <notifications@github.com> wrote:
> 
> Looks good! For most of the files under src/timezone my review was just a diff against upstream and manual inspection, followed by running some of the timezone-related tests. If there's anything else you'd like me to check on my machine, let me know.

Thanks for reviewing! I don’t really know what more to test other than that. The PR pipeline refuse to load/show the failure it reports now so I’m going to rebase it to get a fresh build. If nothing else surface I’ll try to merge this during this week.

**Comment by danielgustafsson** [comment]
When tested in the full pipeline this has given two odd walrep failures, rebased for another run to get to the bottom of what is going on.

**Comment by danielgustafsson** [comment]
Pushed to master following a clean PR pipeline run, thanks for review! Once this has been sitting in master and stabilized a bit it should be backported to 5X_STABLE as well.

**Pull request merged**